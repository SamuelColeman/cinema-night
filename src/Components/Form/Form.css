form {
	height: 100vh;
	margin: 0px;
}

input {
	width: 72vh;
  height: 6vh;
  font-size: 4vh;
  background: transparent;
  color: white;
}