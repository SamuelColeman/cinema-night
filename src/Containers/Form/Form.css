form {
	height: 100vh;
  margin: 0px;
  display: flex;
  flex-direction: column;
  justify-content: center;

}

.form-input {
	width: 33%;
  height: 6vh;
  font-size: 4vh;
  margin-left: auto;
  margin-right: auto;
  margin-top: 0;
  padding-left: 2%;
}

.form-button {
  width: 35.5%;
  height: 6vh;
  margin-left:auto;
  margin-right:auto;
  margin-top:1%;
  background-color: rgba(207,203,202,.5);
  color: white;
}

.form-button:hover {
  background-color:  rgba(48,52,63,.7);
  font-size: 1.2em;
  font-weight:bold;
  cursor:pointer;
  color: white;
}

.form-error {
  color: red;
}

.form-title {
  padding: 0;
}

.form__h1--title {
  color: white;
  margin-bottom: 5%;
  font-size: 2em;
  border-bottom: solid;

}