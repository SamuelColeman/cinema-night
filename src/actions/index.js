export const getMovies = (movies) => ({
  type: 'GET_MOVIES',
  movies
})

export const isLoading = (isLoading) => ({
  type: 'IS_LOADING',
  isLoading
})

export const hasError = (errorMsg) => ({
  type: 'HAS_ERROR',
  errorMsg
})

export const getEmail = (email) => ({
  type: 'GET_EMAIL',
  email
})

export const getPassword = (password) => ({
  type: 'GET_PASSWORD',
  password
})