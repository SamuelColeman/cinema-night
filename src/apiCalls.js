export const currentMovies = async () => {
  const response = await fetch('https://api.themoviedb.org/3/movie/now_playing?api_key=2adea2e47475ecbf6312f332fc8e9ee2');
  const data = await response.json();
  data.results.forEach(movie => {
  	movie.poster_path = `https://image.tmdb.org/t/p/w500${movie.poster_path}`
    movie.backdrop_path = `https://image.tmdb.org/t/p/w500${movie.backdrop_path}`
  })
  return data.results;
}

export const loginVerification = async (info) => { 
  const options = {
    method: 'POST',
    body: JSON.stringify(info),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  const resp = await fetch('http://localhost:3001/api/v1/login', options)
  const data = await resp.json();
  console.log('fetch data-->', data)
  return data
}

export const signUpVerification = async (info) => {
  const options = {
    method: 'POST',
    body: JSON.stringify(info),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  const resp = await fetch('http://localhost:3001/api/v1/users', options)
  const data = await resp.json();
  console.log('fetch data-->', data)
  return data
}

export const addFavourite = async ( id, movie ) => {
  console.log('IN POST!!!', movie, id)
  const options = {
    method: 'POST',
    body: JSON.stringify(movie),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  const resp =  await fetch(`http://localhost:3001/api/v1/users/${id}/moviefavorites`, options);
  const data = await resp.json();
  console.log(data)
  return data
}

export const getFavourites = async ( id ) => {

  const resp =  await fetch(`http://localhost:3001/api/v1/users/${id}/moviefavorites`);
  const data = await resp.json();
  return data
}

export const deleteFavorite = async ( userId, movieId ) => {
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }

  const resp =  await fetch(`http://localhost:3001/api/v1/users/${userId}/moviefavorites/${movieId}`, options);
  const data = await resp.json();
  return data
}
