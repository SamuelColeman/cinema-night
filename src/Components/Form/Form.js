import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
// import { getEmail, getPassword } from '../../actions';
import { login, signUp } from '../../actions'
import MoviesContainer from '../MoviesContainer/MoviesContainer'
import { Route, Link } from 'react-router-dom';
import { loginVerification, signUpVerification, getFavourites } from '../../apiCalls'

class Form extends Component{
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      error: ''
    }
  }

  handleChange = (e) => {
    e.preventDefault()
    this.setState({ [e.target.name]: e.target.value });
    console.log(this.state)
  }

  verifySignIn = async e => {
    e.preventDefault()
    console.log('state', this.state)
    const resp = await loginVerification({
      email: this.state.email,
      password: this.state.password
    })
    if(resp.id) {
      this.props.login({
        name:resp.name,
        id: resp.id,
        isSignedIn: true
      })
      this.showFavourites(resp.id)
    }
      if(resp.error !== undefined) {
        this.setState({error: 'Email and password do not match.'})
      } else {
      this.setState({error: ''})
    }
  }

  showFavourites = async (id) => {
      
      const resp = await getFavourites(id);

      console.log(resp);
  }

  verifySignUp = async e => {
    e.preventDefault()
    const resp = await signUpVerification({
      name: 'pants',
      email: this.state.email,
      password: this.state.password
    })
    if(resp.error !== undefined) {
      this.setState({error: 'Email is taken, try another.'})
    } else {
      this.setState({error: ''})
    }
    if(!resp.email) {
      this.props.signUp({
       name: 'pants',
       email: this.state.email,
       password: this.state.password
      })
    } 
  }

   render() {
     const { email, password, error } = this.state;
    return (
        <form>
            <input placeholder='Email' 
                   type='email'
                   name='email'
                   value={email} 
                   onChange={this.handleChange} />
            <h1>{error}</h1>
            <input placeholder='Password must 8 characters' 
                   type='password'
                   name='password'
                   value={password} 
                   minLength='8' 
                   onChange={this.handleChange} />
            <button onClick={this.verifySignUp}>Sign Up</button>
            <button onClick={this.verifySignIn}>Sign In</button>
            <Link to='/' >
              <button type='button'>Back To Movies</button> 
            </Link>
        </form>
      )
   }
}

export const mapStateToProps = ({ currentUser, users }) => ({
    currentUser,
    users
})

export const mapDispatchToProps = (dispatch) => (
    bindActionCreators({
    login: info => dispatch(login(info)),
    signUp: info  => dispatch(signUp(info))
    }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(Form);