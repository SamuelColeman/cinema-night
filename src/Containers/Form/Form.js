import React, { Component } from 'react';
import './Form.css'; 
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { login, signUp, favouritesList, hasError } from '../../actions'
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types'
import { loginVerification, signUpVerification, getFavourites } from '../../apiCalls'

export class Form extends Component{
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      error: ''
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  verifySignIn = async e => {
    e.preventDefault()
    const { hasError } = this.props;
    const resp = await loginVerification({
      email: this.state.email,
      password: this.state.password
    })
    if(resp.id) {
      this.props.login({
        name:resp.name,
        id: resp.id,
        isSignedIn: true
      })
      this.showFavourites(resp.id)
      hasError(''); 
    }
      if(resp.error !== undefined) {
        this.setState({error: 'Email and password do not match.'})
      } else {
      this.setState({error: ''})
    }
  }

  showFavourites = async (id) => {
      const resp = await getFavourites(id);
      if(resp.favorites) {
        this.props.favouritesList({favorites: resp.favorites })
      }
      if(resp.error !== undefined) {
        this.setState({error: 'Failed to fetch favourites.'})
      } else {
      this.setState({error: ''})
    }
    console.log(resp.favorites)
  }

  verifySignUp = async e => {
    e.preventDefault()
    const resp = await signUpVerification({
      name: 'pants',
      email: this.state.email,
      password: this.state.password
    })
    if(resp.error !== undefined) {
      this.setState({error: 'Email is taken, try another.'})
    } else {
      this.setState({error: ''})
    }
    if(!resp.email) {
      this.props.signUp({
       name: 'pants',
       email: this.state.email,
       password: this.state.password
      })
    } 
  }

   render() {
     const { email, password, error } = this.state;
    return (
        <form>
            <h2>Email:</h2>
            <input placeholder='Email' 
                   type='email'
                   name='email'
                   value={email} 
                   onChange={this.handleChange} />
            <h1>{error}</h1>
            <h2>Password:</h2>
            <input placeholder='Password must 8 characters' 
                   type='password'
                   name='password'
                   value={password} 
                   minLength='8' 
                   onChange={this.handleChange} />
            <button onClick={this.verifySignUp}>Sign Up</button>
            <button onClick={this.verifySignIn}>Sign In</button>
            <Link to='/' >
              <button type='button'>Back To Movies</button> 
            </Link>
        </form>
      )
   }
}

export const mapStateToProps = ({ currentUser }) => ({
    currentUser
})

export const mapDispatchToProps = (dispatch) => (
    bindActionCreators({
    login: info => dispatch(login(info)),
    signUp: info  => dispatch(signUp(info)),
    favouritesList: info => dispatch(favouritesList(info)),
    hasError: info => dispatch(hasError(info))
    }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(Form);

Form.propTypes = {
  currentUser: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
    isSignedIn: PropTypes.bool
  }).isRequired,
  favouritesList: PropTypes.func,
  hasError: PropTypes.func,
  login: PropTypes.func,
  signUp: PropTypes.func,
  email: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  error: PropTypes.string.isRequired
}

Form.defaultProps = {
  currentUser: {},
  email: '',
  password: '',
  error: ''
}