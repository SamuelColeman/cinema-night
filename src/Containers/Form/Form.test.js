import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { Link } from 'react-router-dom';
import { Form, mapStateToProps, mapDispatchToProps } from './Form';
import { login, signUp, favouritesList, hasError } from '../../actions';
import { loginVerification, signUpVerification, getFavourites } from '../../apiCalls';

jest.mock('../../apiCalls');

describe('FormContainer',  () => { 
   
    describe('Form', () => {
       let wrapper; 
       let mockCurrentUser = {
        name: 'Bob',
        id: 1,
        isSignedIn: true
       };
       let mockLogin = mockCurrentUser;
       let mockSignUp = {
          name: 'Bob',
          email: 'bob@gmail.com',
          password: 'bob12345'
       };
       let mockFavouritesList = [
        { 
          id: 456,
          title: 'Harry Potter and the Sorceres Stone',
          poster_path: 'https://pottersworld.com',
          release_date: '10-02-10',
          popularity: 462.91,
          vote_avarage: 3775,
          overview: 'yada yada',
        }
       ];
       let mockError = 'Error';
  

       beforeEach(() => {
           wrapper = shallow (<Form 
                currentUser={mockCurrentUser} 
                login={mockLogin} 
                signUp={mockSignUp} 
                favouritesList={mockFavouritesList} 
                hasError={mockError} />);

          loginVerification.mockImplementation(() => {
              return Promise.resolve(mockLogin);
          });

          signUpVerification.mockImplementation(() => {
              return Promise.resolve(mockSignUp);
          });

          getFavourites.mockImplementation(() => {
              return Promise.resolve(mockCurrentUser.id);
          })
       })

       it('should match snapshot with all information passing through', () => {
           expect(wrapper).toMatchSnapshot();
       })

       it('should update state when verifySignIn is called', () => {
         const mockEvent = {
           preventDefault: jest.fn()
         }

         wrapper.setState({ email: 'bob@gmail.com', password: 'bob1234' });
         wrapper.instance().verifySignIn(mockEvent);
         expect(loginVerification).toHaveBeenCalledWith({ email: 'bob@gmail.com', password: 'bob1234'});
       })

       it('should call loginVerification fetch when verifySignIn is called', () => {
        wrapper.instance().verifySignIn();
        expect(loginVerification).toHaveBeenCalled();
      })

      it('should update state when showFavourites is called', () => {
          let mockMovieId = 4567;
          wrapper.setState({ id: mockMovieId });
          wrapper.instance().showFavourites(mockMovieId);
          expect(getFavourites).toHaveBeenCalledWith(mockMovieId);
      })

      it('should call getFavourites fetch when showFavourites is called', () => {
        wrapper.instance().showFavourites();
        expect(getFavourites).toHaveBeenCalled();
      })

      it('should update state when verifySignUp is called', () => {
        const mockEvent = {
          preventDefault: jest.fn()
        }
        wrapper.setState({ name: 'pants', email: 'bob@gmail.com', password: 'bob1234'});
        wrapper.instance().verifySignUp(mockEvent);
        expect(signUpVerification).toHaveBeenCalledWith({ name: 'pants', email: 'bob@gmail.com', password: 'bob1234'});
      })

      it('should call signUpVerification fetch when verifySignUp is called', () => {
        wrapper.instance().verifySignUp();
        expect(signUpVerification).toHaveBeenCalled();
      })

       it('should update local state of email when handle change is invoked', () => {
           const mockEvent = { 
               target : {
               name: 'email',
               value: 'bob@gmail.com',
             } 
           }

           const expected = 'bob@gmail.com';

           wrapper.instance().handleChange(mockEvent);

           expect(wrapper.state('email')).toEqual(expected);
       })

       it('should update local state of password when handle change is invoked', () => {
         let mockEvent = { 
           target : {
             name: 'password',
             value: 'bobstheword1234',
           } 
         };

         const expected = 'bobstheword1234';
 
         wrapper.instance().handleChange(mockEvent);

         expect(wrapper.state('password')).toEqual(expected);
       })

       it('should call verifySignIn when sign in button is clicked', () => {
           wrapper.instance().verifySignIn = jest.fn();

           const mockEvent = { preventDefault: jest.fn() };

           wrapper.instance().forceUpdate();

           wrapper.find('button').at(1).simulate('click', mockEvent);

           expect(wrapper.instance().verifySignIn).toHaveBeenCalled();
       })

       it('should call verifySignUp when sign up button is clicked', () => {
          wrapper.instance().verifySignUp = jest.fn();

          const mockEvent = { preventDefault: jest.fn() };

          wrapper.instance().forceUpdate();

          wrapper.find('button').at(0).simulate('click', mockEvent);

          expect(wrapper.instance().verifySignUp).toHaveBeenCalled();
    })

    it('should redirect back to main page when button is clicked', () => {
       let wrapper = mount(
           <MemoryRouter initialEntries={['/' ]}>
              <Link to='/' >
                <button type='button'>Back To Movies</button> 
              </Link>
           </MemoryRouter >
       )

       expect(wrapper.find(Link)).toHaveLength(1);
    })
       // test verifySignIn async
       // showFavourites async
       // verifySignUp async
       // snapshot test

    });

    describe('mapStateToProps', () => {
       it('should return a current user when checking state', () => {
          const mockState = {
             currentUser : {
                name : 'Bob',
                id: 1,
                isSignedIn: true
             },
             filter: 'LOGIN_USER'
          }

          const expected = {
            currentUser : {
                name : 'Bob',
                id: 1,
                isSignedIn: true
             }
          }

          const mappedProps = mapStateToProps(mockState);

          expect(mappedProps).toEqual(expected);
       })
    });

    describe('mapDispatchToProps', () => {
      //login
      it.skip('should dispatch login when verifySignIn is called',  () => {
          const mockDispatch = jest.fn();
          const actionToDispatch = login('Bob', 1, false);

          const mappedProps = mapDispatchToProps(mockDispatch);
          mappedProps.verifySignIn();

          expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
      })
      //signup
      //favouritesList
      //hasError
    });
})