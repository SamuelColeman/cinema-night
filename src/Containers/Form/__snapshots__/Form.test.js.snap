// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContainer Form should match snapshot with all information passing through 1`] = `
<form>
  <h2>
    Email:
  </h2>
  <input
    name="email"
    onChange={[Function]}
    placeholder="Email"
    type="email"
    value=""
  />
  <h1 />
  <h2>
    Password:
  </h2>
  <input
    minLength="8"
    name="password"
    onChange={[Function]}
    placeholder="Password must 8 characters"
    type="password"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Sign Up
  </button>
  <button
    onClick={[Function]}
  >
    Sign In
  </button>
  <Link
    to="/"
  >
    <button
      type="button"
    >
      Back To Movies
    </button>
  </Link>
</form>
`;
